go/chat/localizer.go

```diff
@@ -0,0 +1,952 @@
+	defer l.Unlock()
+	res = new(localizerPipelineJob)
+	res.ctx, res.cancelFn = context.WithCancel(BackgroundContext(l.ctx, g))
+	res.retCh = l.retCh
+	res.uid = l.uid
+	res.completed = l.completed
+	res.pending = make([]chat1.Conversation, len(l.pending))
+	res.gateCh = make(chan struct{})
+	copy(res.pending, l.pending)
+	return res
+}
+
+func (l *localizerPipelineJob) closeIfDone() bool {
+	l.Lock()
+	defer l.Unlock()
+	if len(l.pending) == 0 {
+		close(l.retCh)
+		return true
+	}
+	return false
+}
+
+func (l *localizerPipelineJob) getPending() (res []chat1.Conversation) {
+	l.Lock()
+	defer l.Unlock()
+	res = make([]chat1.Conversation, len(l.pending))
+	copy(res, l.pending)
+	return res
+}
+
+func (l *localizerPipelineJob) numPending() int {
+	l.Lock()
+	defer l.Unlock()
+	return len(l.pending)
+}
+
+func (l *localizerPipelineJob) numCompleted() int {
+	l.Lock()
+	defer l.Unlock()
+	return l.completed
+}
+
+func (l *localizerPipelineJob) complete(convID chat1.ConversationID) {
+	l.Lock()
+	defer l.Unlock()
+	for index, j := range l.pending {
+		if j.GetConvID().Eq(convID) {
+			l.completed++
+			l.pending = append(l.pending[:index], l.pending[index+1:]...)
+			return
+		}
+	}
+}
+
+func newLocalizerPipelineJob(ctx context.Context, g *globals.Context, uid gregor1.UID,
+	convs []chat1.Conversation, retCh chan types.AsyncInboxResult) *localizerPipelineJob {
+	return &localizerPipelineJob{
+		ctx:     BackgroundContext(ctx, g),
+		retCh:   retCh,
+		uid:     uid,
+		pending: convs,
+		gateCh:  make(chan struct{}),
+	}
+}
+
+type localizerPipeline struct {
+	globals.Contextified
+	utils.DebugLabeler
+	sync.Mutex
+
+	offline bool
+
+	started   bool
+	stopCh    chan struct{}
+	cancelCh  chan struct{}
+	suspendWg sync.WaitGroup
+	jobQueue  chan *localizerPipelineJob
+
+	// testing
+	useGateCh   bool
+	jobPulledCh chan *localizerPipelineJob
+}
+
+func newLocalizerPipeline(g *globals.Context) *localizerPipeline {
+	return &localizerPipeline{
+		Contextified: globals.NewContextified(g),
+		DebugLabeler: utils.NewDebugLabeler(g.GetLog(), "localizerPipeline", false),
+		stopCh:       make(chan struct{}),
+		cancelCh:     make(chan struct{}),
+	}
+}
+
+func (s *localizerPipeline) Connected() {
+	s.Lock()
+	defer s.Unlock()
+	s.offline = false
+}
+
+func (s *localizerPipeline) Disconnected() {
+	s.Lock()
+	defer s.Unlock()
+	s.offline = true
+}
+
+func (s *localizerPipeline) queue(ctx context.Context, uid gregor1.UID, convs []chat1.Conversation,
+	retCh chan types.AsyncInboxResult) {
+	defer s.Trace(ctx, func() error { return nil }, "queue")()
+	s.Lock()
+	defer s.Unlock()
+	job := newLocalizerPipelineJob(ctx, s.G(), uid, convs, retCh)
+	job.ctx, job.cancelFn = context.WithCancel(BackgroundContext(ctx, s.G()))
+	select {
+	case s.jobQueue <- job:
+	default:
+		s.Debug(ctx, "queue: dropped job, queue full")
+	}
+	s.Debug(ctx, "queue: len: %d", len(s.jobQueue))
+}
+
+func (s *localizerPipeline) clearQueue() {
+	s.jobQueue = make(chan *localizerPipelineJob, 500)
```

Comment:
can you make this configurable and drop it to 1-2 in tests to see if any bad behavior is triggered?

Reply:
I'll just make `queue` block if the buffer is full, think it is ok.